pkg/steps/template.go

```diff
@@ -48,6 +57,61 @@ type templateExecutionStep struct {
 	subTests []*junit.TestCase
 }
 
+func injectResourcesToTemplatePods(template *templateapi.Template, resources api.ResourceConfiguration) error {
+	if resources != nil {
+		containerResources, err := resourcesFor(resources.RequirementsForTemplateStep(template.Name))
+		if err != nil {
+			return fmt.Errorf("unable to calculate resources for %s: %s", template.Name, err)
+		}
+
+		corev1Scheme := runtime.NewScheme()
+		utilruntime.Must(coreapi.AddToScheme(corev1Scheme))
+		corev1Codec := serializer.NewCodecFactory(corev1Scheme).LegacyCodec(coreapi.SchemeGroupVersion)
+
+		for i, object := range template.Objects {
```

Comment:
This is still really duplicated with the other call-site. Why not  do this:

```go
func operateOnTemplatePods(template *templateapi.Template, func(pod *coreapi.Pod)) error {}
```

The fact that the code in the other file does crazy JSON marshalling is not important -- we should refactor it to mutate the Pod in Go code and only marshal it at the end to make the raw JSON, just like you are doing here for this. It is not important that we iterate over the list twice. O(2N) === O(N).

Reply:
@smarterclayton could you please give us the reasoning behind the JSON-based workflow you implemented to inject the artifacts container?

Reply:
this is why I left the PR as WIP.  I will address this. Any more insights are welcome.